internal/app/connectconformance/results.go

```diff
@@ -317,65 +317,69 @@ func headerValsToString(vals []string) string {
 		actualPayload := actual[i]
 		expectedPayload := expected[i]
 		if !bytes.Equal(actualPayload.Data, expectedPayload.Data) {
 			errs = append(errs, fmt.Errorf("response #%d: expecting data %x, got %x", i+1, expectedPayload.Data, actualPayload.Data))
 		}
-		actualReq := actualPayload.GetRequestInfo()
-		expectedReq := expectedPayload.GetRequestInfo()
-
-		if i == 0 { //nolint:nestif
-			// Validate headers, timeout, and query params. We only need to do this once, for first payload.
-			errs = append(errs, checkHeaders("request headers", expectedReq.GetRequestHeaders(), actualReq.GetRequestHeaders())...)
-			if expectedReq != nil && expectedReq.TimeoutMs != nil {
-				if actualReq == nil || actualReq.TimeoutMs == nil {
-					errs = append(errs, fmt.Errorf("server did not echo back a timeout but one was expected (%d ms)", expectedReq.GetTimeoutMs()))
-				} else {
-					max := expectedReq.GetTimeoutMs()
-					min := max - timeoutCheckGracePeriodMillis
-					if min < 0 {
-						min = 0
-					}
-					if actualReq.GetTimeoutMs() > max || actualReq.GetTimeoutMs() < min {
-						errs = append(errs, fmt.Errorf("server echoed back a timeout (%d ms) that did not match expected (%d ms)", actualReq.GetTimeoutMs(), expectedReq.GetTimeoutMs()))
-					}
-				}
-			} else if actualReq != nil && actualReq.TimeoutMs != nil {
-				errs = append(errs, fmt.Errorf("server echoed back a timeout (%d ms) but none was expected", actualReq.GetTimeoutMs()))
-			}
-			if len(expectedReq.GetConnectGetInfo().GetQueryParams()) > 0 && len(actualReq.GetConnectGetInfo().GetQueryParams()) > 0 {
-				errs = append(errs, checkHeaders("request query params", expectedReq.GetConnectGetInfo().GetQueryParams(), actualReq.GetConnectGetInfo().GetQueryParams())...)
-			}
-		}
 
-		if len(actualReq.GetRequests()) != len(expectedReq.GetRequests()) {
-			errs = append(errs, fmt.Errorf("response #%d: expecting %d request messages to be described but instead got %d", i+1, len(expectedReq.GetRequests()), len(actualReq.GetRequests())))
-		}
-		for i := 0; i < len(actualReq.GetRequests()) && i < len(expectedReq.GetRequests()); i++ {
-			reqNum++
-			actualMsg, err := anypb.UnmarshalNew(actualReq.GetRequests()[i], proto.UnmarshalOptions{})
-			if err != nil {
-				errs = append(errs, fmt.Errorf("request #%d: failed to unmarshal actual message: %w", reqNum, err))
-				continue
-			}
-			expectedMsg, err := anypb.UnmarshalNew(expectedReq.GetRequests()[i], proto.UnmarshalOptions{})
-			if err != nil {
-				errs = append(errs, fmt.Errorf("request #%d: failed to unmarshal expected message: %w", reqNum, err))
-				continue
-			}
-			diff := cmp.Diff(expectedMsg, actualMsg, protocmp.Transform())
-			if diff != "" {
-				errs = append(errs, fmt.Errorf("request #%d: did not survive round-trip: - wanted, + got\n%s", reqNum, diff))
-			}
-		}
+		errs = append(errs, checkRequestInfo(expectedPayload.GetRequestInfo(), actualPayload.GetRequestInfo(), i)...)
```

Comment:
Instead of passing `i`, I think you can check it here:
```go
if i == 0 {
    // Validate headers, timeout, and query params. We only need to do this once, for first response.
    errs = append(errs, checkRequestInfo(expectedPayload.GetRequestInfo(), actualPayload.GetRequestInfo())...)
}
```
That way you can omit the hard-coded zero in the other call site.

Reply:
The only problem is that value is also used on line 347 if request lengths don't match:

` errs = append(errs, fmt.Errorf("response #%d: expecting %d request messages to be described but instead got %d", respNum+1, len(expected.GetRequests()), len(actual.GetRequests())))`

Reply:
I just removed the number from that `fmt.Errorf` statement. It probably doesn't provide a ton of info when weighed against having to pass the number in as a param.